core/shared/src/main/scala/org/http4s/NonNegative.scala

```diff
@@ -0,0 +1,56 @@
+package org.http4s
+
+import java.io.Serializable
+import scala.math.BigInt
+
+sealed abstract class NonNegative extends Serializable {
+  import NonNegative._
+
+  def toInt: Option[Int]
+  def toLong: Option[Long]
+  def toBigInt: BigInt
+
+  override def hashCode = toBigInt.hashCode
+
+  override def equals(that: Any) = (this, that) match {
+    case (LongNonNegative(x), LongNonNegative(y)) => x == y
+    case (BigIntNonNegative(x), BigIntNonNegative(y)) => x == y
+    case (x, y: NonNegative) => x.toBigInt == y.toBigInt
+    case (_, _) => false
+  }
+}
+
+object NonNegative {
```

Comment:
Also, we might add the `cats.kernel.Hash` instance, though it's useless for the moment

Reply:
And Eq and Order and LowerBound and Monoid.  There are a bunch.  This was just a silly PoC.